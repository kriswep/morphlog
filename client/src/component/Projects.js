import React from 'react';
import gql from 'graphql-tag';
import { graphql, compose } from 'react-apollo';
import { Link } from 'react-router-dom';
import styled from 'styled-components';
import { Route, Redirect } from 'react-router';

import media from '../styles/media';
import Project from './Project';
import Input from './Input';

const SidebarContainer = styled.section`
  grid-area: sidebar;
  padding: 0.25rem;
  ${media.m`
    padding 0.75rem;
  `};
`;

const ProjectsContainer = styled.ul`
  list-style-type: none;
  padding: 0;
  margin: 0;
`;

const ProjectLink = styled(Link)`
  color: ${props => props.theme.darkShades};
`;

const initialState = {
  name: '',
};

const ProjectWrapper = props => {
  const projectId = props;
  return <Project {...props} />;
};

class Projects extends React.Component {
  state = initialState;

  dispatch = (e, v, x) => {
    if (e.target.name && e.target.name) {
      const newState = {};
      newState[e.target.name] = e.target.value;
      this.setState(newState);
    }
  };

  addProject = e => {
    this.props
      .addProjectMutation({
        variables: {
          name: this.state.name,
        },
        refetchQueries: [
          {
            query: PROJECTS_QUERY,
          },
        ],
      })
      .then(({ data }) => {
        console.log('got data', data);
      })
      .catch(error => {
        console.log('there was an error sending the query', error);
      })
      .finally(() => {
        this.setState(initialState);
      });
  };

  render() {
    const projects = this.props.projectsQuery.projects;

    let projectId = this.props.match.params.projectId;
    if (
      !this.props.match.params.projectId &&
      projects &&
      projects.length >= 0 &&
      projects[0].id
    ) {
      // no project open, but we have projects: render project comp
      projectId = projects[0].id;
    }
    return [
      projectId && <Project projectId={projectId} />,
      <SidebarContainer>
        <h2>Projects</h2>
        <Input
          name="name"
          label="new project name"
          value={this.state.name}
          type="text"
          placeholder="awesome project"
          onChange={this.dispatch}
        />
        <button onClick={this.addProject}>add</button>
        {projects && (
          <ProjectsContainer>
            {projects.map(project => (
              <li key={project.id}>
                <ProjectLink to={`/project/${project.id}`}>
                  {project.name}
                </ProjectLink>
              </li>
            ))}
          </ProjectsContainer>
        )}
      </SidebarContainer>,
    ];
  }
}

const PROJECTS_QUERY = gql`
  query projects {
    projects {
      id
      name
    }
  }
`;
const ADD_PROJECT_MUTATION = gql`
  mutation addProject($name: String!) {
    createProject(name: $name) {
      id
      name
    }
  }
`;

export default compose(
  graphql(PROJECTS_QUERY, { name: 'projectsQuery' }),
  graphql(ADD_PROJECT_MUTATION, { name: 'addProjectMutation' }),
)(Projects);
