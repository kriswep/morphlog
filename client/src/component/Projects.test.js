/* globals test expect jest window */
import React from 'react';
import { ApolloProvider } from 'react-apollo';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import createClient from '../../utils/apolloMocks';
// import Project from './Project';
// import { Link } from 'react-router-dom';

import Projects from './Projects';

jest.mock('./Project', () => () => <div>MockedProject</div>);
jest.mock('react-router-dom', () => {
  Link: () => <a>MockedLink</a>; // eslint-disable-line
});

const mocks = {
  Query: () => ({
    projects: () => [
      {
        id: '1234',
        text: 'MyProject',
        createdAt: '2018-01-21T20:45:06.000Z',
        updatedAt: '2018-01-21T20:45:06.000Z',
      },
      {
        id: '5678',
        text: 'MyOtherProject',
        createdAt: '2018-01-21T20:45:06.000Z',
        updatedAt: '2018-01-21T20:45:06.000Z',
      },
    ],
  }),

  // Mutation: () => ...
};

const client = createClient(mocks);

test('Projects renders correctly', async () => {
  const match = {
    params: {
      projectId: 'cjcp94wxp025801100npb28yg',
    },
  };
  const wrapper = mount(
    <ApolloProvider client={client}>
      <Projects match={match} />
    </ApolloProvider>,
  );
  await new Promise(res => window.setTimeout(res, 1));
  wrapper.setProps({ projectId: 'bar' }); // poke it to rerender...
  expect(toJson(wrapper.find('[data-test="projects"]'))).toMatchSnapshot();
  expect(toJson(wrapper.find('[data-test="newProject"]'))).toMatchSnapshot();
});
