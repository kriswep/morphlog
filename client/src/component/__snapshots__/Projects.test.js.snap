// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects renders correctly 1`] = `
Array [
  <styled.ul
    data-test="projects"
>
    <ul
        className="sc-bZQynM eCIGDO"
        data-test="projects"
    >
        <li
            key="1234"
        >
            <Styled(Link)
                to="/project/1234"
            >
                <Link
                    className="sc-gzVnrw ZyJST"
                    to="/project/1234"
                >
                    <a>
                        MockedLink
                    </a>
                </Link>
            </Styled(Link)>
        </li>
        <li
            key="5678"
        >
            <Styled(Link)
                to="/project/5678"
            >
                <Link
                    className="sc-gzVnrw ZyJST"
                    to="/project/5678"
                >
                    <a>
                        MockedLink
                    </a>
                </Link>
            </Styled(Link)>
        </li>
    </ul>
</styled.ul>,
  <ul
    className="sc-bZQynM eCIGDO"
    data-test="projects"
>
    <li
        key="1234"
    >
        <Styled(Link)
            to="/project/1234"
        >
            <Link
                className="sc-gzVnrw ZyJST"
                to="/project/1234"
            >
                <a>
                    MockedLink
                </a>
            </Link>
        </Styled(Link)>
    </li>
    <li
        key="5678"
    >
        <Styled(Link)
            to="/project/5678"
        >
            <Link
                className="sc-gzVnrw ZyJST"
                to="/project/5678"
            >
                <a>
                    MockedLink
                </a>
            </Link>
        </Styled(Link)>
    </li>
</ul>,
]
`;

exports[`Projects renders correctly 2`] = `
<div
  data-test="newProject"
>
  <MyInput
    label="New project name"
    name="name"
    onChange={[Function]}
    placeholder="awesome project"
    type="text"
    value=""
  >
    <styled.div>
      <div
        className="sc-bdVaJa bPGLpq"
      >
        <styled.label
          for="name"
        >
          <label
            className="sc-bwzfXH hgqFEP"
          >
            New project name
          </label>
        </styled.label>
        <styled.input
          name="name"
          onChange={[Function]}
          placeholder="awesome project"
          type="text"
          value=""
        >
          <input
            className="sc-htpNat hHFhwJ"
            name="name"
            onChange={[Function]}
            placeholder="awesome project"
            type="text"
            value=""
          />
        </styled.input>
      </div>
    </styled.div>
  </MyInput>
  <Button
    animated={true}
    as="button"
    color="teal"
    onClick={[Function]}
    role="button"
  >
    <button
      className="ui teal animated button"
      onClick={[Function]}
      role="button"
    >
      <ButtonContent
        visible={true}
      >
        <div
          className="visible content"
        >
          Add
        </div>
      </ButtonContent>
      <ButtonContent
        hidden={true}
      >
        <div
          className="hidden content"
        >
          <Icon
            as="i"
            name="right arrow"
          >
            <i
              aria-hidden="true"
              className="right arrow icon"
            />
          </Icon>
        </div>
      </ButtonContent>
    </button>
  </Button>
</div>
`;
